File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_java_labs_and_solutions/observer/lab/Meeting.java
// Meeting.java

public class Meeting
{
	public Meeting(int aMonth, int aDay, int aYear, int aHour, int aMinute, String aLocation)
	{
		// do not call set_XXX() here, as they will call update().
		myMonth    = aMonth;
		myDay      = aDay;
		myYear     = aYear;
		myHour     = aHour;
		myMinute   = aMinute;
		myLocation = aLocation;

		myManagers   = new DpVector<Manager>();
		myDevelopers = new DpVector<Developer>();
	}

	public void add(Manager aManager)
	{
		System.out.println("Meeting.add(aManager)");

		if (aManager == null)
		{
			throw new DpException("void add(Manager aManager) -- aManager == null.");
		}

		myManagers.add(aManager);

		aManager.add(this);
	}

	public void add(Developer aDeveloper)
	{
		System.out.println("Meeting.add(aDeveloper)");

		if (aDeveloper == null)
		{
			throw new DpException("void add(Developer aDeveloper) -- aDeveloper == null.");
		}

		myDevelopers.add(aDeveloper);

		aDeveloper.add(this);
	}

	public DpIterator<Manager> getManagers()
	{
		return myManagers.makeIterator();
	}

	public DpIterator<Developer> getDevelopers()
	{
		return myDevelopers.makeIterator();
	}

	private void update()
	{
		System.out.println("Meeting.update()");

		{
			DpIterator<Manager> aIterator = getManagers();
			while (!aIterator.isDone())
			{
				Manager aManager = aIterator.getCurrent();
				aManager.update(this);
				aIterator.advance();
			}
		}

		{
			DpIterator<Developer> aIterator = getDevelopers();
			while (!aIterator.isDone())
			{
				Developer aDeveloper = aIterator.getCurrent();
				aDeveloper.update(this);
				aIterator.advance();
			}
		}
	}

	public void cancel()
	{
		System.out.println("Meeting.cancel()");

		{
			DpIterator<Manager> aIterator = getManagers();
			while (!aIterator.isDone())
			{
				Manager aManager = aIterator.getCurrent();
				aManager.remove(this);
				aIterator.advance();
			}
		}

		{
			DpIterator<Developer> aIterator = getDevelopers();
			while (!aIterator.isDone())
			{
				Developer aDeveloper = aIterator.getCurrent();
				aDeveloper.remove(this);
				aIterator.advance();
			}
		}
	}

	public String toString()
	{
		return
		""
		+ myMonth
		+ "/"
		+ myDay
		+ "/"
		+ myYear
		+ "."
		+ myHour
		+ ":"
		+ myMinute
		+ " at "
		+ myLocation;
	}

	public void setMonth(int aMonth)
	{
		myMonth = aMonth;
		update();
	}

	public void setDay(int aDay)
	{
		myDay = aDay;
		update();
	}

	public void setYear(int aYear)
	{
		myYear = aYear;
		update();
	}

	public void setHour(int aHour)
	{
		myHour = aHour;
		update();
	}

	public void setMinute(int aMinute)
	{
		myMinute = aMinute;
		update();
	}

	public void setLocation(String aLocation)
	{
		myLocation = aLocation;
		update();
	}

	private int myMonth;
	private int myDay;
	private int myYear;
	private int myHour;
	private int myMinute;
	private String myLocation;

	private DpVector<Manager>   myManagers;
	private DpVector<Developer> myDevelopers;
}

File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_cpp_labs_and_solutions/observer/lab/meeting.cpp
// meeting.cpp

#include <iostream>
#include "developer.h"
#include "dp_exception.h"
#include "dp_iterator.h"
#include "dp_vector.h"
#include "manager.h"
#include "meeting.h"
#include "to_string.h"

meeting::meeting(int a_month, int a_day, int a_year, int a_hour, int a_minute, const std::string & a_location)
{
	// do not call set_XXX() here, as they will call update().
	my_month    = a_month;
	my_day      = a_day;
	my_year     = a_year;
	my_hour     = a_hour;
	my_minute   = a_minute;
	my_location = a_location;
}

void meeting::add(manager * a_manager)
{
	std::cout << "meeting.add(a_manager)" << std::endl;

	if (a_manager == nullptr)
	{
		throw dp_exception("void meeting::add(manager * a_manager) -- a_manager == nullptr.");
	}

	my_managers.add(a_manager);

	a_manager->add(this);
}

void meeting::add(developer * a_developer)
{
	std::cout << "meeting.add(a_developer)" << std::endl;

	if (a_developer == nullptr)
	{
		throw dp_exception("void meeting::add(developer * a_developer) -- a_developer == nullptr.");
	}

	my_developers.add(a_developer);

	a_developer->add(this);
}

dp_iterator<manager *> * meeting::get_managers()
{
	return my_managers.make_iterator();
}

dp_iterator<developer *> * meeting::get_developers()
{
	return my_developers.make_iterator();
}

void meeting::update()
{
	std::cout << "meeting.update()" << std::endl;

	{
		dp_iterator<manager *> * a_iterator = get_managers();
		while (!a_iterator->is_done())
		{
			manager * a_manager = a_iterator->get_current();
			a_manager->update(this);
			a_iterator->advance();
		}
	}

	{
		dp_iterator<developer *> * a_iterator = get_developers();
		while (!a_iterator->is_done())
		{
			developer * a_developer = a_iterator->get_current();
			a_developer->update(this);
			a_iterator->advance();
		}
	}
}

void meeting::cancel()
{
	std::cout << "meeting.cancel()" << std::endl;

	{
		dp_iterator<manager *> * a_iterator = get_managers();
		while (!a_iterator->is_done())
		{
			manager * a_manager = a_iterator->get_current();
			a_manager->remove(this);
			a_iterator->advance();
		}
	}

	{
		dp_iterator<developer *> * a_iterator = get_developers();
		while (!a_iterator->is_done())
		{
			developer * a_developer = a_iterator->get_current();
			a_developer->remove(this);
			a_iterator->advance();
		}
	}
}

std::string meeting::to_string() const
{
	return
	""
	+ ::to_string(my_month)
	+ "/"
	+ ::to_string(my_day)
	+ "/"
	+ ::to_string(my_year)
	+ "."
	+ ::to_string(my_hour)
	+ ":"
	+ ::to_string(my_minute)
	+ " at "
	+ my_location;
}

void meeting::set_month(int a_month)
{
	my_month = a_month;
	update();
}

void meeting::set_day(int a_day)
{
	my_day = a_day;
	update();
}

void meeting::set_year(int a_year)
{
	my_year = a_year;
	update();
}

void meeting::set_hour(int a_hour)
{
	my_hour = a_hour;
	update();
}

void meeting::set_minute(int a_minute)
{
	my_minute = a_minute;
	update();
}

void meeting::set_location(const std::string & a_location)
{
	my_location = a_location;
	update();
}
